### PR TITLE
Fix/6067: Require Re-evaluations of QA and EM

### LINKED ISSUE #6067: Require Re-evaluations of QA and EM data issues
There is no logic in place to require re-evaluations of already evaluated QA and EM data when dependent changes to a MP occur.

Currently able to evaluate QA and EM data with open windows, make a change to the MP and not evaluate it, and continue with the QA and EM submissions.

The change to the MP should force the MP, QA, and EM data to be re-evaluated prior to any submissions.

Evaluating the MP does set the QA back to needs evaluation, but not the EM data.

Evaluating the QA data also does not set the EM data back to needs evaluation.

The logic in 1.0 to require re-evaluations of QA and EM data with open windows needs to be reviewed and carried over to 2.0.

QA and EM is dependent on MP.

EM is also dependent on QA.

### Acceptance Tests

1. Importing or changing a MP should require evaluation of the MP.
2. The evaluation of a MP that has been updated or is required for resubmission should force re-evaluation of the following (when they are in the workspace) prior to their submission. 
   1. EM where the EM Needs Evaluation equals 'N', and the associated EM Submission Access (ESA) Submission Availability Code equals 'REQUIRE' or 'GRANTED'.
      1. Should set **Needs Evaluation Flag** to 'Y', and clear the **Check Session Id** and the associated **Check Session** and **Check Log** rows.
      2. Should clear calculated columns (begin with "CALC_") from any tables associated with the EM report.
      3. Should clear the APPLICABLE_BIAS_ADJ_FACTOR for the associated Derived and Monitor Hourly Value rows associated with the EM report.
      4. Should clear calculated columns (begin with "CALC_") and the Error Codes column from the Emission View tables for rows association with the EM report.
   2. QA Tests (QAT) for locations in the MP where the QAT Needs Evaluation equals 'N'.
      1.  Should set **Needs Evaluation Flag** to 'Y', and clear the **Check Session Id** and the associated **Check Session** and **Check Log** rows.
      2. Should clear calculated columns (begin with "CALC_") from any tables associated with the test.
   3. QA Cert Events (QCE) for locations in the MP where the QCE Needs Evaluation equals 'N', and the QCE has a Submission Availability Code of 'REQUIRE' or the QCE has been updated.
      1. Should set **Needs Evaluation Flag** to 'Y', clear the **Check Session Id** and the associated **Check Session** and **Check Log** rows, and set the **Update Date** to the current time.
   4. Test Extensions and Exemptions (TEE) for locations in the MP where the TEE Needs Evaluation equals 'N', and the TEE has a Submission Availability Code of 'REQUIRE' or the TEE has been updated.
      1. Should set **Needs Evaluation Flag** to 'Y', clear the **Check Session Id** and the associated **Check Session** and **Check Log** rows, and set the **Update Date** to the current time.
3. The evaluation of a QA Test (QAT) that does not have a QA_SUPP_DATA row (never evaluated or submitted) or has a QA_SUPP_DATA row that indicates the test can be submitted, should force re-evaluation of the following (when they are in the workspace) prior to their submission. 
   1. QA reports that depend on the evaluated QA report based on one of the following: 
      1. QA Update Dependencies
         1. Update F2LREF (test type) when a RATA (test type) is evaluated and the following are true:
            - F2LREF and RATA MON_SYS_ID are the same.
            - F2LREF NEEDS_EVAL_FLG equals 'N'.
            - F2LREF time is on or after the RATA time.
            - F2LREF has not been submitted or is required to submit or granted permission to submit.
         2. Update F2LCHK when a RATA or F2LREF is evaluated and the following are true:
            - F2LCHK MON_SYS_ID equals the evaluated test's MON_SYS_ID.
            - F2LCHK NEEDS_EVAL_FLG equals 'N'.
            - F2LCHK quarter is on or after the quarter of the evaluated test END_DATE.
            - F2LCHK has not been submitted or is required to submit or granted permission to submit.
         3. Update FF2LTST when a FF2LBAS is evaluated and the following are true:
            - FF2LTST MON_SYS_ID equals the evaluated test's MON_SYS_ID.
            - FF2LTST NEEDS_EVAL_FLG equals 'N'.
            - FF2LTST quarter is on or after the quarter of the FF2LBAS END_DATE.
            - FF2LTST has not been submitted or is required to submit or granted permission to submit.
         4. Update FF2LBAS when a FFACC or FFACCTT is evaluated and the following are true:
            - FF2LBAS ACCURACY_TEST_NUMBER (FUEL_FLOW_TO_LOAD_BASELINE) equals the evaluated test's TEST_NUM.
            - FF2LBAS NEEDS_EVAL_FLG equals 'N'.
            - FF2LBAS  has not been submitted or is required to submit or granted permission to submit.
         5. Update FF2LBAS when a PEI is evaluated and the following are true:
            - FF2LBAS PEI_TEST_NUMBER (FUEL_FLOW_TO_LOAD_BASELINE) equals the PEI TEST_NUM.
            - FF2LBAS NEEDS_EVAL_FLG equals 'N'.
            - FF2LBAS has not been submitted or is required to submit or granted permission to submit.
      2. QA Update Actions
         1. Reference the MP handling for QA above, from **2.ii.a** through **2.ii.b**.
   2. EM reports prior to their submission, when the EM report is in the workspace, its Needs Evaluation equals 'N', the associated EM Submission Access (ESA) Submission Availability Code equals 'REQUIRE' or 'GRANTED', and the EM Report quarter is on or after the QAT Quarter or End Date quarter.
      1. Reference the MP handling for EM above, from **2.i.a** through **2.i.d**.
4. The evaluation of a QA Cert Event (QCE) that has been created, updated or required to resubmit should force re-evaluation of EM reports prior to their submission, when the EM report is in the workspace, its Needs Evaluation equals 'N', the associated EM Submission Access (ESA) Submission Availability Code equals 'REQUIRE' or 'GRANTED', and the EM Report quarter is on or after the QCE QA Cert Event Date quarter.
   1. Reference the MP handling for EM above, from **2.i.a** through **2.i.d**.
5. The evaluation of a Test Exemption and Extension (TEE) that has been created, updated or required to resubmit should force re-evaluation of EM reports prior to their submission, when the EM report is in the workspace, its Needs Evaluation equals 'N', the associated EM Submission Access (ESA) Submission Availability Code equals 'REQUIRE' or 'GRANTED', and the EM Report quarter is on or after the TEE Quarter.
   1. Reference the MP handling for EM above, from **2.i.a** through **2.i.d**.
6. The evaluation of an Emission Report (EM) should result in the following:
   1. Regenerate the workspace Emission Views for the EM Report by calling REFRESH_EMISSIONS_VIEWS (CAMDECMPSWKS).
   2. Should set **Needs Evaluation Flag** to 'Y', and clear the **Check Session Id** and the associated **Check Session** and **Check Log** rows when the evaluation returned HOURGEN-8-C.
      - Determine by linking from the EMISSION_EVALUATION row for the MON_PLAN_ID and RPT_PERIOD_ID to CHECK_LOG on CHK_SESSION_ID, CHECK_CATALOG_RESULT on CHECK_CATALOG_RESULT_ID, and CHECK_CATALOG on CHECK_CATALOG_ID.  CHECK_CATALOG will contain check types and check numbers, and CHECK_CATALOG result will contain check results.
   3. The evaluation of an Emission Report (EM) that has been required or given permission to resubmit should force re-evaluation of EM reports prior to their submission, when the EM report is in the workspace and:
      1. The EM Report contains a monitor location in the evaluated emission report.
      3. The EM Report's quarter is on or after the Evaluated EM Report's quarter.
      4. The associated EM Submission Access (ESA) Submission Availability Code equals 'REQUIRE' or 'GRANTED'.

### Developer Notes:

1. **UPDATE_ECMPS_STATUS_FOR_MP_EVALUATION** (camdecmpswks): The section to handle resetting emissions after a MP evaluation is commented out and has not been implemented.  This accounts for why MP evaluations correctly handle QA, but do not handle EM.
2. **UPDATE_ECMPS_STATUS_FOR_QA_EVALUATION** (camdecmpswks): The section to handle resetting emissions after a QA evaluation is commented out and has not been implemented.  This accounts for why QA evaluations do not correctly handle EM.  This section should also handle resetting test to evaluation needed when they are dependent on an evaluated test.
3. **UPDATE_ECMPS_STATUS_FOR_QCE_EVALUATION** (camdecmpswks): The section to handle resetting emissions after a QCE evaluation is commented out and has not been implemented.
4.  **UPDATE_ECMPS_STATUS_FOR_TEE_EVALUATION** (camdecmpswks): The section to handle resetting emissions after a TEE evaluation is commented out and has not been implemented.
5. **UPDATE_ECMPS_STATUS_FOR_EM_EVALUATION**(camdecmpswks): The calls to recreate EM Grid Data and to conditionally set MP Needs Evaluation was not implemented.  This section should also sets MP to needs evaluation when the EM evaluation returns HOURGEN-8-C.
6. Although the UPDATE_ECMPS_STATUS Evaluation processes for MP, QAT, QCE, TEE, and EM where implemented, the other processes were not implemented, and I believe that CVP did not plan to implement them as database procedures.  The resetting of NEEDS_EVAL_FLG that occurs for those processes **needs to occur in application code**.  Handle by [#6291](https://github.com/US-EPA-CAMD/easey-ui/issues/6291)

#### Zip of UPDATE_ECMPS_STATUS and related ECMPS 1.0 scripts.
[Update ECMPS Status SPs.zip](https://github.com/user-attachments/files/15570686/Update.ECMPS.Status.SPs.zip)

### PR DESCRIPTION
**Pull request by ayang2017**

### CODE REVIEW

File: camdecmpswks/functions/ay_update_ecmps_status_for_em_evaluation.sql
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_em_evaluation.sql:1]: A large portion of the code for the 'EM Evaluation' process in the UPDATE_ECMPS_STATUS procedure from ECMPS 1.0 (SQL Server) is missing from this version.  I extracted the 'EM Evaluation' code and included it below.  This code only includes the CHECK_SESSION delete and the EMISSION_EVALUATION update.

```
SET @SUBMITTABLE = 'N'

SELECT @SUBMITTABLE = 'Y'	
    FROM EM_SUBMISSION_ACCESS
    WHERE MON_PLAN_ID = @V_ID_KEY_OR_LIST AND
          RPT_PERIOD_ID = @V_PERIOD_ID AND
          SUBMISSION_AVAILABILITY_CD IN ('GRANTED', 'REQUIRE');

-- delete old evaluation data from check logs
DELETE FROM CHECK_SESSION 
    WHERE PROCESS_CD = 'HOURLY' AND
        MON_PLAN_ID = @V_ID_KEY_OR_LIST AND
        RPT_PERIOD_ID = @V_PERIOD_ID AND
        CHK_SESSION_ID <> @V_OTHER_FIELD

UPDATE EMISSION_EVALUATION
    SET NEEDS_EVAL_FLG	= 'N',
        CHK_SESSION_ID	= @V_OTHER_FIELD	
    WHERE MON_PLAN_ID = @V_ID_KEY_OR_LIST AND
        RPT_PERIOD_ID = @V_PERIOD_ID

-- Recreate the View Emission grid data
EXEC Client_EM_Grid_Data @V_MON_PLAN_ID = @V_ID_KEY_OR_LIST, @V_RPT_PERIOD_ID = @V_PERIOD_ID, @V_RESULT = @V_RESULT OUTPUT, @V_ERROR_MSG = @V_ERROR_MSG OUTPUT;

--
-- Deleting Calculated data failed, bail (PJR)
--
IF @V_RESULT = 'F'
    RETURN;

-- Update Monitoring Plan Needs Evaluation, if needed
EXEC UES.UpdateMpNeedsEvalForEmEvalIfNeeded @vChkSessionId = @V_OTHER_FIELD, @vMonPlanId = @V_ID_KEY_OR_LIST, @vResult = @V_RESULT OUTPUT, @vErrorMsg = @V_ERROR_MSG OUTPUT;

IF @V_RESULT = 'F'
    RETURN;


IF @SUBMITTABLE = 'Y' 
BEGIN
    INSERT INTO @tmpEmissionsStatus SELECT DISTINCT E.MON_PLAN_ID, E.RPT_PERIOD_ID
        FROM EMISSION_EVALUATION E,
        EM_SUBMISSION_ACCESS ESA,
        MONITOR_PLAN_LOCATION M, 
        REPORTING_PERIOD R,
        (SELECT MON_LOC_ID, EE.MON_PLAN_ID, EE.RPT_PERIOD_ID,
                CALENDAR_YEAR, QUARTER 
            FROM EMISSION_EVALUATION EE
                INNER JOIN REPORTING_PERIOD RP ON EE.RPT_PERIOD_ID = RP.RPT_PERIOD_ID
                INNER JOIN MONITOR_PLAN_LOCATION MPL ON EE.MON_PLAN_ID = MPL.MON_PLAN_ID) T
            WHERE E.MON_PLAN_ID = M.MON_PLAN_ID AND 
                E.RPT_PERIOD_ID = R.RPT_PERIOD_ID AND
                E.MON_PLAN_ID = ESA.MON_PLAN_ID AND 
                E.RPT_PERIOD_ID = ESA.RPT_PERIOD_ID AND
                ESA.SUBMISSION_AVAILABILITY_CD IN ('GRANTED','REQUIRE') AND
                M.MON_LOC_ID = T.MON_LOC_ID AND
                (R.CALENDAR_YEAR > T.CALENDAR_YEAR OR 
                (R.CALENDAR_YEAR = T.CALENDAR_YEAR AND R.QUARTER > T.QUARTER)) AND
                T.MON_PLAN_ID = @V_ID_KEY_OR_LIST  AND
                T.RPT_PERIOD_ID = cast(@V_PERIOD_ID as varchar);


    DECLARE EM_CSR CURSOR local FOR
        SELECT MON_PLAN_ID, RPT_PERIOD_ID
        FROM @tmpEmissionsStatus;

    OPEN EM_CSR;
    FETCH EM_CSR INTO @V_MON_PLAN_ID, @V_RPT_PERIOD_ID;

    WHILE (@@fetch_status = 0) 
    BEGIN
        EXEC dbo.DEL_CALCULATED_EMISSIONS_DATA @V_MON_PLAN_ID = @V_MON_PLAN_ID, @V_RPT_PERIOD_ID = @V_RPT_PERIOD_ID, @V_RESULT = @V_RESULT OUTPUT, @V_ERROR_MSG = @V_ERROR_MSG OUTPUT;
        IF @V_RESULT = 'F'
            BREAK;
        FETCH EM_CSR INTO @V_MON_PLAN_ID, @V_RPT_PERIOD_ID;
    END

    CLOSE EM_CSR;
    DEALLOCATE EM_CSR;

    --
    -- Deleting Calculated data failed, bail (PJR)
    --
    IF @V_RESULT = 'F'
        RETURN;
END

```

File: camdecmpswks/functions/ay_update_ecmps_status_for_mp_evaluation.sql [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+				     from camdecmpswks.delete_calculated_em_data_from_workspace(V_MON_PLAN_ID, V_RPT_PERIOD_ID);	
+					IF result = 'F' then
+					 return next;
+				    end if;	
+				 end loop;
+				CLOSE EM_CSR;
+				DEALLOCATE EM_CSR;	 
+
+  end if;  --vSubmittable = 'Y' 
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_mp_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_mp_evaluation.sql [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+exception when others then
+    get stacked diagnostics error_msg := message_text;
+    result = 'F';
+    error_msg :='From update_ecmps_status_for_mp_evaluation' ||' '|| error_msg;
+	
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_mp_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_em_evaluation.sql [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+				
+	
+  return next;
+  
+exception when others then
+    get stacked diagnostics error_msg:= message_text;
+    result = 'F';
+     error_msg :='From update_ecmps_status_for_em_evaluation' ||' '|| error_msg;
+	 
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_em_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_em_evaluation.sql [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                CHK_SESSION_ID <> vchk_session_id;
+            
+    UPDATE camdecmpswks.EMISSION_EVALUATION
+            SET NEEDS_EVAL_FLG    = 'N',
+                CHK_SESSION_ID    = vchk_session_id    
+            WHERE MON_PLAN_ID = vmon_plan_id AND
+                RPT_PERIOD_ID = vperiod_id;
+				
+	
+  return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_em_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_qa_evaluation.sql [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+				   from camdecmpswks.delete_calculated_em_data_from_workspace(V_MON_PLAN_ID, V_RPT_PERIOD_ID);	
+					IF result = 'F' then
+					 return next;
+				    end if;
+                end loop;
+				CLOSE EM_CSR;
+				DEALLOCATE EM_CSR;  			
+	 
+   end if;
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_qa_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_qa_evaluation.sql [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+exception when others then
+    get stacked diagnostics error_msg := message_text;
+    result = 'F';
+    error_msg :='From update_ecmps_status_for_qa_evaluation ' ||' '|| error_msg;
+	
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_qa_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_qce_evaluation.sql [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+				   select * into result, error_msg 
+				   from camdecmpswks.delete_calculated_em_data_from_workspace(V_MON_PLAN_ID, V_RPT_PERIOD_ID);	
+					IF result = 'F' then
+					 return next;
+				    end if;
+                end loop;
+				CLOSE EM_CSR;
+				DEALLOCATE EM_CSR;  
+	  end if;
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_qce_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_qce_evaluation.sql [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+exception when others then
+    get stacked diagnostics error_msg := message_text;
+    result = 'F';
+    error_msg :='From update_ecmps_status_for_qce_evaluation' ||' '|| error_msg;
+	
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_qce_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_tee_evaluation.sql [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+				   select * into result, error_msg 
+				     from camdecmpswks.delete_calculated_em_data_from_workspace(V_MON_PLAN_ID, V_RPT_PERIOD_ID);	
+					IF result = 'F' then
+					 return next;
+				    end if;	
+				 end loop;
+				CLOSE EM_CSR;
+				DEALLOCATE EM_CSR;	
+	 end if; 
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_tee_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/ay_update_ecmps_status_for_tee_evaluation.sql [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+exception when others then
+    get stacked diagnostics error_msg := message_text;
+    result = 'F';
+    error_msg :='From update_ecmps_status_for_tee_evaluation' ||' '|| error_msg;
+	
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/ay_update_ecmps_status_for_tee_evaluation.sql]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/delete_calculated_em_data_from_workspace.sql [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	-- Delete rows for DM_EMISSIONS and its child tables
+   	DELETE FROM camdecmps.DM_EMISSIONS
+		WHERE MON_PLAN_ID = vmonplan_id
+		  and RPT_PERIOD_ID=vrptperiod_id;
+	
+	--Remove error codes and calculated values from pre-rendered View Emissions tables.
+	 select * into result, error_msg 
+	   from camdecmpswks.emissions_grid_remove_eval(vmonplan_id, vrptperiod_id );	
+	
+	 return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/delete_calculated_em_data_from_workspace.sql:130]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/delete_calculated_em_data_from_workspace.sql [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+exception when others then
+    get stacked diagnostics error_msg := message_text;
+    result = 'F'; 
+	error_msg :='From delete_calculated_em_data_from_workspace '||' '|| message_text;
+	
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/delete_calculated_em_data_from_workspace.sql:137]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/emissions_grid_remove_eval.sql [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+		SET CALC_ADJ_NOX_RATE = NULL,
+			CALC_CO2_MASS_RATE = NULL,
+			CALC_HI_RATE = NULL,
+			CALC_NOX_MASS = NULL,
+			CALC_SO2_MASS_RATE = NULL,
+			ERROR_CODES = NULL
+		where MON_PLAN_ID =vmonplan_id
+		  and RPT_PERIOD_ID = vrptperiod_id;
+
+ return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/emissions_grid_remove_eval.sql:171]: Should this be a 'return' , not a 'return next'?

File: camdecmpswks/functions/emissions_grid_remove_eval.sql [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+exception when others then
+    get stacked diagnostics error_msg := message_text;
+    result = 'F';  
+	error_msg :='From emissions_grid_remove_eval '||' '|| message_text;
+	
+   return next;
```
Comment (djw4erg) [inline camdecmpswks/functions/emissions_grid_remove_eval.sql:178]: Should this be a 'return' , not a 'return next'?

**Review by djw4erg — CHANGES_REQUESTED** [review]
All the comments about 'return next' are just questions, not a request for a change.  **The use of 'return next' might be correct** since 'return next' is new to me.

The comment on ay_**update_ecmps_status_for_em_evaluation.sql** is a request for change.

**Comment by ayang2017** [comment]
There are two functions called by em_evaluation that need create/update:
Client_EM_Grid_Data;
UES.UpdateMpNeedsEvalForEmEvalIfNeeded.

**Comment by ayang2017** [comment]
Updated scripts as requested.

**Pull request merged**